mat64/lq.go

```diff
@@ -9,185 +9,204 @@ import (
-// in the least squares solution of non-square systems of simultaneous linear equations.
-// This will fail if LQIsFullRank() returns false. The matrix a is overwritten by the
-// decomposition.
-func LQ(a *Dense) LQFactor {
-	// Initialize.
+// The LQ decomposition is a factorization of the matrix A such that A = L * Q.
+// The matrix Q is an orthonormal n×n matrix, and L is an m×n upper triangular matrix.
+// L and Q can be extracted from the LFromLQ and QFromLQ methods on Dense.
+func (lq *LQ) Factorize(a Matrix) {
 	m, n := a.Dims()
 	if m > n {
 		panic(ErrShape)
 	}
-
-	lq := *a
-
-	lDiag := make([]float64, m)
-	projs := NewVector(m, nil)
-
-	// Main loop.
-	for k := 0; k < m; k++ {
-		hh := lq.RawRowView(k)[k:]
-		norm := blas64.Nrm2(len(hh), blas64.Vector{Inc: 1, Data: hh})
-		lDiag[k] = norm
-
-		if norm != 0 {
-			hhNorm := (norm * math.Sqrt(1-hh[0]/norm))
-			if hhNorm == 0 {
-				hh[0] = 0
-			} else {
-				// Form k-th Householder vector.
-				s := 1 / hhNorm
-				hh[0] -= norm
-				blas64.Scal(len(hh), s, blas64.Vector{Inc: 1, Data: hh})
-
-				// Apply transformation to remaining columns.
-				if k < m-1 {
-					a = lq.View(k+1, k, m-k-1, n-k).(*Dense)
-					projs = projs.ViewVec(0, m-k-1)
-					projs.MulVec(a, NewVector(len(hh), hh))
-
-					for j := 0; j < m-k-1; j++ {
-						dst := a.RawRowView(j)
-						blas64.Axpy(len(dst), -projs.at(j),
-							blas64.Vector{Inc: 1, Data: hh},
-							blas64.Vector{Inc: 1, Data: dst},
-						)
-					}
-				}
-			}
-		}
+	k := min(m, n)
+	if lq.lq == nil {
+		lq.lq = &Dense{}
 	}
-	*a = lq
-
-	return LQFactor{a, lDiag}
+	lq.lq.Clone(a)
+	work := make([]float64, 1)
+	lq.tau = make([]float64, k)
+	lapack64.Gelqf(lq.lq.mat, lq.tau, work, -1)
+	work = make([]float64, int(work[0]))
+	lapack64.Gelqf(lq.lq.mat, lq.tau, work, len(work))
 }
 
-// IsFullRank returns whether the L matrix and hence a has full rank.
-func (f LQFactor) IsFullRank() bool {
-	for _, v := range f.lDiag {
-		if v == 0 {
-			return false
-		}
+// TODO(btracey): Add in the "Reduced" forms for extracting the m×m orthogonal
+// and upper triangular matrices.
+
+// LFromLQ extracts the m×n lower trapezoidal matrix from a LQ decomposition.
+func (m *Dense) LFromLQ(lq *LQ) {
+	r, c := lq.lq.Dims()
+	m.reuseAs(r, c)
+
+	// Disguise the LQ as a lower triangular
+	t := &TriDense{
+		blas64.Triangular{
+			N:      r,
+			Stride: lq.lq.mat.Stride,
+			Data:   lq.lq.mat.Data,
+			Uplo:   blas.Lower,
+			Diag:   blas.NonUnit,
+		},
 	}
-	return true
-}
+	m.Copy(t)
 
-// L returns the lower triangular factor for the LQ decomposition.
-func (f LQFactor) L() *Dense {
-	lq, lDiag := f.LQ, f.lDiag
-	m, _ := lq.Dims()
-	l := NewDense(m, m, nil)
-	for i, v := range lDiag {
-		for j := 0; j < m; j++ {
-			if i < j {
-				l.set(j, i, lq.at(j, i))
-			} else if i == j {
-				l.set(j, i, v)
-			}
-		}
+	if r == c {
+		return
+	}
+	// Zero right of the triangular.
+	for i := 0; i < r; i++ {
```

Comment:
These are more specialized than those functions. They zero outside the leading triangle.